client/db/bolt/db.go

```diff
@@ -1624,6 +1624,129 @@ func (idx *timeIndexNewest) add(t uint64, k []byte, b *bbolt.Bucket) {
 	})
 }
 
+// DeleteInactiveOrders deletes orders that no longer needed for normal
+// operations. Optionally accepts a time to delete orders with a later time
+// stamp.
+func (db *BoltDB) DeleteInactiveOrders(olderThan *time.Time) (deletedOrders []*dexdb.MetaOrder, finished bool, err error) {
+	start := time.Now()
+	const batchSize = 1000
+	var olderThanB []byte
+	if olderThan != nil {
+		olderThanB = uint64Bytes(encode.UnixMilliU(*olderThan))
+	} else {
+		olderThanB = uint64Bytes(encode.UnixMilliU(start))
+	}
+	deletedOrders = make([]*dexdb.MetaOrder, 0, batchSize)
+	return deletedOrders, finished, db.Update(func(tx *bbolt.Tx) error {
+		// Some archived orders may still be needed for active matches.
+		// Put those order id's in a map to prevent deletion.
+		amb := tx.Bucket(activeMatchesBucket)
+		activeMatches := make(map[order.OrderID]struct{})
+		amb.ForEach(func(k, _ []byte) error {
+			mBkt := amb.Bucket(k)
+			if mBkt == nil {
+				return fmt.Errorf("match %x bucket is not a bucket", k)
+			}
+
+			oidB := mBkt.Get(orderIDKey)
+			var oid order.OrderID
+			copy(oid[:], oidB)
+			activeMatches[oid] = struct{}{}
+			return nil
+		})
+
+		// Run through the archived order bucket storing id's keys for
+		// returning the order data and deletion later.
+		archivedOB := tx.Bucket(archivedOrdersBucket)
+		if archivedOB == nil {
+			return fmt.Errorf("failed to open %s bucket", string(archivedOrdersBucket))
+		}
+		// Retrieve one more than we will delete in order to guage if
+		// this is the last order to delete.
+		oneOverSize := batchSize + 1
+		filter := func(k []byte, oBkt *bbolt.Bucket) bool {
+			var oid order.OrderID
+			copy(oid[:], k)
+			if order.OrderStatus(intCoder.Uint16(oBkt.Get(statusKey))).IsActive() {
+				db.log.Warnf("active order %v found in inactive bucket", oid)
+				return false
+			}
+			if _, has := activeMatches[oid]; has {
+				return false
+			}
+			timeB := oBkt.Get(updateTimeKey)
+			return bytes.Compare(timeB, olderThanB) <= 0
+		}
+		trios := newestBuckets([]*bbolt.Bucket{archivedOB}, oneOverSize, updateTimeKey, filter)
+
+		// Ignore the last order if it exists. Otherwise there are no
+		// more orders to delete.
+		if len(trios) == oneOverSize {
+			trios = trios[:batchSize]
+		} else {
+			finished = true
+		}
+		for _, trio := range trios {
+			o, err := decodeOrderBucket(trio.k, trio.b)
+			if err != nil {
+				return fmt.Errorf("failed to decode order bucket: %v", err)
+			}
+			deletedOrders = append(deletedOrders, o)
+			if err := archivedOB.DeleteBucket(trio.k); err != nil {
```

Comment:
Large bolt transactions always have me concerned about memory.  One would think that deleting buckets wouldn't balloon memory, but... wish we could be more certain about this first.

Reply:
It should be only deleting the batch size per transaction. That can be adjusted.